### ISSUE TITLE
build: post_ci failure should not affect on_success being called

### ISSUE THREAD

**Issue by cratz** [Shippable/support#2315]
`on_success` section is not called after tests passed:

https://app.shippable.com/runs/56d6abcbac215a0c00be8dc2

**Comment by manishas** [comment]
@cratz i looked at the build link you posted and there was an error in your build_ci step, so on_failure was called.

Can you point me to a build where build_ci succeeded and on_success was not called?

**Comment by cratz** [comment]
@manishas 
Thanks for checking. 
The error happened in the `build_post_ci` section and not in the `build_ci` section. After checking the documentation ("Depending on the whether your ci section is successful or not, the on_success or on_failure sections will be executed.") I thought `on_success` is triggered after `build_ci` (`ci` in the yaml file) so it would be possible to create or move build artifacts with this step and use it in the `build_post_ci` section for building the new image. 
If the `on_success` is only triggered after all other steps of `build` it is almost useless for us.

**Comment by manishas** [comment]
makes sense. thanks for clarifying.... we'll work on a fix.

**Comment by cratz** [comment]
@manishas 
Thanks again.
Don't get me wrong I thought that because of the documentation. May be it is intended to work like this and other users using it in a different way.
An event like `on_ci_success` or similar would allow to group the build section logically and keep existing events supported.

Best

**Comment by manishas** [comment]
makes sense :)

I plan to check in with the team and make sure we're handling this correctly.

**Comment by pquentin** [comment]
Any progress on this?

**Comment by rageshkrishna** [comment]
@cratz @pquentin Can you please shed some light on the kind of workflow that you are trying to get to? I'd like to first nderstand exactly what you're trying to accomplish so I can figure out what needs to change. Thanks!

**Comment by pquentin** [comment]
It's a documentation issue. We want to make sure we're using each step correctly and we want to know exactly what's going to happen when a given step fails.

**Comment by cratz** [comment]
@rageshkrishna 
@pquentin is right after reading the documentation I thought `on_success` is triggered after the `ci` step so I tried to use this section for generating build artifacts and other stuff which should happen after the tests passed. Than I recognized that this event is only triggered when no error happen in the whole `build` section. I just moved all steps from `on_success` to `ci` and everything works as expected. 

Basically I can't see any reason for using `on_success` since I can just add the commands to the end of the `ci` section and it will behave the same way. Only `on_failure` makes sense and `on_success` was / is just confusing.
